src/streaming/thread.pool.cpp

```diff
@@ -0,0 +1,96 @@
+#include "thread.pool.hh"
+
+zarr::ThreadPool::ThreadPool(unsigned int n_threads,
+                             std::function<void(const std::string&)> err)
+  : error_handler_{ err }
+  , is_accepting_jobs_{ true }
+{
+    n_threads = std::clamp(
+      n_threads, 1u, std::max(std::thread::hardware_concurrency(), 1u));
```

Comment:
A clearer way to write this might be:

```cpp
const auto max_threads = std::max(std::thread::hardware_concurrency(), 1u);
n_threads = std::clamp(n_threads, 1u, max_threads);
```